dtm0/dtx.c

```diff
@@ -0,0 +1,276 @@
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+
+/**
+ * @addtogroup dtm
+ *
+ * @{
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM
+#include "dtm0/dtx.h"
+#include "lib/assert.h" /* M0_PRE */
+#include "lib/memory.h" /* M0_ALLOC */
+#include "lib/errno.h"  /* ENOMEM */
+#include "lib/trace.h"  /* M0_ERR */
+#include "dtm0/service.h" /* m0_dtm0_service */
+#include "reqh/reqh.h" /* reqh2confc */
+#include "conf/helpers.h" /* proc2srv */
+
+static struct m0_sm_state_descr dtx_states[] = {
+	[M0_DDS_INIT] = {
+		.sd_flags     = M0_SDF_INITIAL | M0_SDF_FINAL,
+		.sd_name      = "init",
+		.sd_allowed   = M0_BITS(M0_DDS_INPROGRESS),
+	},
+	[M0_DDS_INPROGRESS] = {
+		.sd_name      = "dtx-executed",
+		.sd_allowed   = M0_BITS(M0_DDS_EXECUTED, M0_DDS_FAILED),
+	},
+	[M0_DDS_EXECUTED] = {
+		.sd_name      = "dtx-persistent",
+		.sd_allowed   = M0_BITS(M0_DDS_PERSISTENT),
+	},
+	[M0_DDS_PERSISTENT] = {
+		.sd_name      = "dtx-persistent",
+		.sd_allowed   = M0_BITS(M0_DDS_STABLE),
+	},
+	[M0_DDS_STABLE] = {
+		.sd_name      = "dtx-stable",
+		.sd_allowed   = M0_BITS(M0_DDS_DONE),
+	},
+	[M0_DDS_DONE] = {
+		.sd_name      = "done",
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+	[M0_DDS_FAILED] = {
+		.sd_name      = "dtx-failed",
+		.sd_flags     = M0_SDF_TERMINAL | M0_SDF_FAILURE
+	}
+};
+
+static struct m0_sm_trans_descr dtx_trans[] = {
+	{ "populated",  M0_DDS_INIT,       M0_DDS_INPROGRESS },
+	{ "executed",   M0_DDS_INPROGRESS, M0_DDS_EXECUTED   },
+	{ "exec-fail",  M0_DDS_INPROGRESS, M0_DDS_FAILED     },
+	{ "persistent", M0_DDS_EXECUTED,   M0_DDS_PERSISTENT },
+	{ "stable",     M0_DDS_PERSISTENT, M0_DDS_STABLE     },
+	{ "prune",      M0_DDS_STABLE,     M0_DDS_DONE       }
+};
+
+static struct m0_sm_conf dtx_sm_conf = {
+	.scf_name      = "dtm0dtx",
+	.scf_nr_states = ARRAY_SIZE(dtx_states),
+	.scf_state     = dtx_states,
+	.scf_trans_nr  = ARRAY_SIZE(dtx_trans),
+	.scf_trans     = dtx_trans,
+};
+
+static int m0_dtm0_dtx_init(struct m0_dtm0_service *svc,
+			    struct m0_sm_group     *sm_group,
+			    struct m0_dtm0_dtx     **out)
+{
+	struct m0_dtm0_dtx *dtx;
+
+	M0_ALLOC_PTR(dtx);
+	if (dtx == NULL)
+		return M0_ERR(-ENOMEM);
+
+	m0_sm_conf_init(&dtx_sm_conf);
+	m0_sm_init(&dtx->dd_sm, &dtx_sm_conf, M0_DDS_INIT, sm_group);
+	dtx->dd_dtms = svc;
+	dtx->dd_ancient_dtx.tx_dtx = dtx;
+	*out = dtx;
+	return 0;
+}
+
+static void m0_dtm0_dtx_fini(struct m0_dtm0_dtx *dtx)
+{
+	m0_sm_fini(&dtx->dd_sm);
+	m0_free(dtx);
+}
+
+static int m0_dtm0_dtx_prepare(struct m0_dtm0_dtx *dtx)
+{
+	int               rc;
+	struct m0_dtm0_ts now;
+
+	rc = m0_dtm0_clk_src_now(&dtx->dd_dtms->dos_clk_src, &now);
+	if (rc != 0)
+		return rc;
+
+	dtx->dd_txd.dtd_id = (struct m0_dtm0_tid) {
+		.dti_ts  = now,
+		.dti_fid = dtx->dd_dtms->dos_generic.rs_service_fid,
+	};
+
+	M0_POST(m0_dtm0_tid__invariant(&dtx->dd_txd.dtd_id));
+	return rc;
+}
```

Comment:
it can be simplified a bit but it cannot be as simple as you described. In your example, the post condition is checked before the timestamp component of the TID is filled. This post condition must ensure that the invariant is held before exiting the scope.